cmd/dmverity-vhd/main.go

```diff
@@ -0,0 +1,251 @@
+			}
+			if verbose {
+				fmt.Fprintf(os.Stdout, "Layer #%d, layer hash: %s\n", layerNumber, diffID.String())
+			}
+
+			rc, err := layer.Uncompressed()
+			if err != nil {
+				return errors.Wrapf(err, "failed to uncompress layer %s", diffID.String())
+			}
+
+			vhdPath := filepath.Join(ctx.String(outputDirFlag), diffID.Hex+".vhd")
+			out, err := os.Create(vhdPath)
+			if err != nil {
+				return errors.Wrapf(err, "failed to create layer vhd %s", vhdPath)
+			}
+
+			if verbose {
+				fmt.Fprintf(os.Stdout, "converting tar to layer VHD\n")
+			}
+			opts := []tar2ext4.Option{
+				tar2ext4.ConvertWhiteout,
+				tar2ext4.MaximumDiskSize(maxVHDSize),
+				tar2ext4.AppendVhdFooter,
+				tar2ext4.AppendDMVerity,
+			}
+			if err := tar2ext4.Convert(rc, out, opts...); err != nil {
+				return errors.Wrap(err, "failed to convert tar to ext4")
+			}
+			fmt.Fprintf(os.Stdout, "Layer %d: %s\n", layerNumber, vhdPath)
+		}
+		return nil
+	},
+}
+
+var rootHashVHDCommand = cli.Command{
+	Name:  "roothash",
+	Usage: "compute root hashes for each LCOW layer VHD",
+	Flags: []cli.Flag{
+		cli.StringFlag{
+			Name:     imageFlag + ",i",
+			Usage:    "Required: container image reference",
+			Required: true,
+		},
+		cli.StringFlag{
+			Name:  usernameFlag + ",u",
+			Usage: "Optional: custom registry username",
+		},
+		cli.StringFlag{
+			Name:  passwordFlag + ",p",
+			Usage: "Optional: custom registry password",
+		},
+	},
+	Action: func(ctx *cli.Context) error {
+		verbose := ctx.GlobalBool(verboseFlag)
+
+		layers, err := fetchImageLayers(ctx)
+		if err != nil {
+			return errors.Wrap(err, "failed to fetch image layers")
+		}
+
+		if verbose {
+			fmt.Fprintf(os.Stdout, "%d layers found\n", len(layers))
+		}
+
+		tmpFile, err := ioutil.TempFile("", "")
+		if err != nil {
+			return errors.Wrap(err, "failed to create temporary file")
+		}
+		defer os.Remove(tmpFile.Name())
+
+		for layerNumber, layer := range layers {
+			diffID, err := layer.DiffID()
+			if err != nil {
+				return errors.Wrap(err, "failed to read layer diff")
+			}
+			if verbose {
+				fmt.Fprintf(os.Stdout, "Layer %d.\nUncompressed layer hash: %s\n", layerNumber, diffID.String())
+			}
+
+			rc, err := layer.Uncompressed()
+			if err != nil {
+				return errors.Wrapf(err, "failed to uncompress layer %s", diffID.String())
+			}
+
+			opts := []tar2ext4.Option{
+				tar2ext4.ConvertWhiteout,
+				tar2ext4.MaximumDiskSize(maxVHDSize),
+			}
+
+			if _, err := tmpFile.Seek(0, io.SeekStart); err != nil {
+				return errors.Wrapf(err, "failed seek start on temp file when processing layer %d", layerNumber)
+			}
+			if err := tmpFile.Truncate(0); err != nil {
+				return errors.Wrapf(err, "failed truncate temp file when processing layer %d", layerNumber)
+			}
+
+			if err := tar2ext4.Convert(rc, tmpFile, opts...); err != nil {
+				return errors.Wrap(err, "failed to convert tar to ext4")
+			}
+
+			data, err := os.ReadFile(tmpFile.Name())
```

Comment:
I was confused as to how the ci was building as this was added to the stdlib (copied from the io/ioutil package) in 1.16 but our ci states that it's using 1.15. I found this in the ci logs
```
Setup go stable version spec ^1.15.0
Found in cache @ C:\hostedtoolcache\windows\go\1.16.2\x64
Added go to the path
Successfully setup go version ^1.15.0
go version go1.16.2 windows/amd64 
```

Reply:
For reference: https://golang.org/pkg/io/ioutil/
"As of Go 1.16, the same functionality is now provided by package io or package os, and those implementations should be preferred in new code. See the specific function documentation for details."

Reply:
yes, you're right. I should move back to `ioutil`. what was the reason we bumped the version in ci?

Reply:
I didn't even realize we were using a caret in the go semver for our ci, so it'll use whatever the latest go release >= 1.15.0 (but less than 2.0.0) is which is why it's pulling in 1.16.2.

https://nodesource.com/blog/semver-tilde-and-caret/#:~:text=Caret%3A%20Flexible%20Minor%20and%20Patch&text=Any%20minor%20or%20patch%20version,%2C%20a%20range%20of%20~1.2.

Reply:
(and thus has the stdlib change)